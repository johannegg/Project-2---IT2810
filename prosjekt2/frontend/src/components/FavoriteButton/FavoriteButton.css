.favoriteButton {
	background-color: transparent;
	font-size: 30px;
}

@media (max-width: 650px) {
    .favoriteButton {
		font-size: 25px;
    }
}

@media (max-width: 450px) {
    .favoriteButton {
		font-size: 20px;
    }
}
