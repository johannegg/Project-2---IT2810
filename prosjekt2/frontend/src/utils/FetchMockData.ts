export interface Song {
	id: number;
	title: string;
	artist: string;
	year: number;
	views: number;
	lyrics: string;
	genre: string;
}

export const fetchSongs = async (): Promise<Song[]> => {
	try {
		const response = await fetch("/mockdata.json");
		if (!response.ok) {
			throw new Error("Failed to fetch songs");
		}
		const data: Song[] = await response.json();
		return data;
	} catch (error) {
		console.error("Error fetching songs:", error);
		throw error;
	}
};
