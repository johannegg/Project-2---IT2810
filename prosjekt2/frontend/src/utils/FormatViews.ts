export const formatViews = (views: number): string => {
	return new Intl.NumberFormat("en", { notation: "compact", compactDisplay: "short" }).format(
		views,
	);
};
