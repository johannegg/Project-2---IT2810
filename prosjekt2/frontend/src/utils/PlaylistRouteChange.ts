import { PlaylistData } from "../pages/Playlists/Playlists";
import { NavigateFunction } from "react-router-dom";

export const routeChange = (playlist: PlaylistData, navigate: NavigateFunction) => {
  const path = `/playlist/${playlist.name.toLowerCase().replace(/ /g, "-")}`;
  navigate(path, { state: playlist });
};
