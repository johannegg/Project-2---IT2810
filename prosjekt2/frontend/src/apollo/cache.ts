import { InMemoryCache, makeVar } from "@apollo/client";
import { SongData } from "../utils/types/SongTypes";
import { PlaylistData } from "../pages/Playlists/Playlists";

// Initialize sessionStorage and localStorage if values are not already set
const initializeStorage = (key: string, defaultValue: string) => {
  if (!sessionStorage.getItem(key)) {
    sessionStorage.setItem(key, defaultValue);
  }
};

const initializeLocalStorage = (key: string, defaultValue: string) => {
  if (!localStorage.getItem(key)) {
    localStorage.setItem(key, defaultValue);
  }
};

// Initialize values in storage
initializeStorage("minViews", "0");
initializeStorage("maxViews", "1000000");
initializeStorage("selectedGenres", "[]");
initializeStorage("sortOption", "views_desc");
initializeStorage("homeSearchTerm", "");
initializeStorage("favoritesSearchTerm", "");
initializeLocalStorage("favoriteSongs", "[]");
initializeLocalStorage("playlists", "[]");

// Load saved values from storage
const savedGenres = JSON.parse(sessionStorage.getItem("selectedGenres") || "[]");
const savedMinViews = Number(sessionStorage.getItem("minViews"));
const savedMaxViews = Number(sessionStorage.getItem("maxViews"));
const savedSortOption = sessionStorage.getItem("sortOption") || "views_desc";
const savedHomeSearchTerm = sessionStorage.getItem("homeSearchTerm") || "";
const savedFavoritesSearchTerm = sessionStorage.getItem("favoritesSearchTerm") || "";
const savedFavoriteSongs = JSON.parse(localStorage.getItem("favoriteSongs") || "[]");
const savedPlaylists = JSON.parse(localStorage.getItem("playlists") || "[]");

// Create reactive variables
export const genreFilterVar = makeVar<string[]>(savedGenres);
export const sortOptionVar = makeVar<string>(savedSortOption);
export const minViewsVar = makeVar<number>(savedMinViews);
export const maxViewsVar = makeVar<number>(savedMaxViews);
export const homeSearchTermVar = makeVar<string>(savedHomeSearchTerm);
export const favoritesSearchTermVar = makeVar<string>(savedFavoritesSearchTerm);
export const favoriteSongsVar = makeVar<SongData[]>(savedFavoriteSongs);
export const playlistsVar = makeVar<PlaylistData[]>(savedPlaylists);  
export const songDataVar = makeVar<SongData[]>([]);

// Configure cache with reactive variables
const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        genreFilter: {
          read() {
            return genreFilterVar();
          },
        },
        sortOption: {
          read() {
            return sortOptionVar();
          },
        },
        minViews: {
          read() {
            return minViewsVar();
          },
        },
        maxViews: {
          read() {
            return maxViewsVar();
          },
        },
        homeSearchTerm: {
          read() {
            return homeSearchTermVar();
          },
        },
        favoritesSearchTerm: {
          read() {
            return favoritesSearchTermVar();
          },
        },
        favoriteSongs: {
          read() {
            return favoriteSongsVar();
          },
        },
        playlists: {
          read() {
            return playlistsVar();
          },
        },
        songData: {
          read() {
            return songDataVar();
          },
        },
      },
    },
  },
});

// Sync reactive variables with sessionStorage and localStorage
homeSearchTermVar.onNextChange((newHomeSearchTerm) => {
  sessionStorage.setItem("homeSearchTerm", newHomeSearchTerm);
});

favoritesSearchTermVar.onNextChange((newFavoritesSearchTerm) => {
  sessionStorage.setItem("favoritesSearchTerm", newFavoritesSearchTerm);
});

genreFilterVar.onNextChange((newGenres) => {
  sessionStorage.setItem("selectedGenres", JSON.stringify(newGenres));
});

minViewsVar.onNextChange((newMinViews) => {
  sessionStorage.setItem("minViews", String(newMinViews));
});

maxViewsVar.onNextChange((newMaxViews) => {
  sessionStorage.setItem("maxViews", String(newMaxViews));
});

sortOptionVar.onNextChange((newSortOption) => {
  sessionStorage.setItem("sortOption", newSortOption);
});

favoriteSongsVar.onNextChange((newFavorites) => {
  localStorage.setItem("favoriteSongs", JSON.stringify(newFavorites));
});

playlistsVar.onNextChange((newPlaylists) => {
  localStorage.setItem("playlists", JSON.stringify(newPlaylists));
  window.dispatchEvent(new Event("storage")); 
});

export default cache;
