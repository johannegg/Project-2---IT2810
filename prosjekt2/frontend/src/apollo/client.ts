import { ApolloClient } from "@apollo/client";
import cache from "./cache";

const client = new ApolloClient({
	uri: "http://localhost:3001/", // Use this to run locally, for tests
	//uri: "http://it2810-12.idi.ntnu.no:3001", // Use this to run on virtual machine
	cache: cache,
});

export default client;
