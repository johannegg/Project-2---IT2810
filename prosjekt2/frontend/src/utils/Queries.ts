import { gql } from "@apollo/client";

export const GET_SONGS = gql`
	query GetSongs(
		$skip: Int
		$limit: Int
		$genres: [String]
		$sortBy: SortBy
		$searchTerm: String
		$minViews: Int
		$maxViews: Int
	) {
		songs(
			skip: $skip
			limit: $limit
			genres: $genres
			sortBy: $sortBy
			searchTerm: $searchTerm
			minViews: $minViews
			maxViews: $maxViews
		) {
			id
			title
			views
			year
			lyrics
			artist {
				name
				id
			}
			genre {
				name
			}
		}
	}
`;

export const GET_SONG_COUNT = gql`
  query GetSongCount(
    $genres: [String]
    $searchTerm: String
    $minViews: Int
    $maxViews: Int
  ) {
    songCount(
      genres: $genres
      searchTerm: $searchTerm
      minViews: $minViews
      maxViews: $maxViews
    )
  }
`;

export const GET_GENRE_COUNTS = gql`
  query GetGenreCounts(
    $searchTerm: String
    $minViews: Int
    $maxViews: Int
    $genres: [String]
  ) {
    genreCounts(
      searchTerm: $searchTerm
      minViews: $minViews
      maxViews: $maxViews
      genres: $genres
    ) {
      name
      count
    }
  }
`;

export const CREATE_USER = gql`
	mutation CreateUser($username: String!) {
		createUser(username: $username) {
			id
			username
			favoriteSongs {
				id
				title
				views
				year
				lyrics
				artist {
					name
				}
				genre {
					name
				}
			}
			playlists {
				id
				name
				songs {
					id
					title
					views
					year
					lyrics
					artist {
						name
						id
					}
					genre {
						name
					}
				}
			}
		}
	}
`;

export const DELETE_USER = gql`
	mutation DeleteUser($username: String!) {
		deleteUser(username: $username)
	}
`;

export const ADD_FAVORITE_SONG = gql`
	mutation AddFavoriteSong($username: String!, $songId: ID!) {
		addFavoriteSong(username: $username, songId: $songId)
	}
`;

export const REMOVE_FAVORITE_SONG = gql`
	mutation RemoveFavoriteSong($username: String!, $songId: ID!) {
		removeFavoriteSong(username: $username, songId: $songId)
	}
`;

export const CREATE_PLAYLIST = gql`
	mutation CreatePlaylist(
		$username: String!
		$name: String!
		$backgroundcolor: String!
		$icon: String!
	) {
		createPlaylist(
			username: $username
			name: $name
			backgroundcolor: $backgroundcolor
			icon: $icon
		) {
			id
			name
			backgroundcolor
			icon
			songs {
				id
				title
			}
		}
	}
`;

export const DELETE_PLAYLIST = gql`
	mutation DeletePlaylist($username: String!, $playlistId: ID!) {
		deletePlaylist(username: $username, playlistId: $playlistId)
	}
`;

export const ADD_SONG_TO_PLAYLIST = gql`
	mutation AddSongToPlaylist($username: String!, $playlistId: ID!, $songId: ID!) {
		addSongToPlaylist(username: $username, playlistId: $playlistId, songId: $songId) {
			id
			name
			backgroundcolor
			icon
			songs {
				id
				title
			}
		}
	}
`;

export const REMOVE_SONG_FROM_PLAYLIST = gql`
	mutation RemoveSongFromPlaylist($username: String!, $playlistId: ID!, $songId: ID!) {
		RemoveSongFromPlaylist(username: $username, playlistId: $playlistId, songId: $songId) {
			id
			name
			backgroundcolor
			icon
			songs {
				id
				title
			}
		}
	}
`;
