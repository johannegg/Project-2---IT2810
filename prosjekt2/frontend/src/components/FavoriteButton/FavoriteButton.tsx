import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faHeart as heartRegular } from "@fortawesome/free-regular-svg-icons";
import { faHeart as heartSolid } from "@fortawesome/free-solid-svg-icons";
import "./FavoriteButton.css";
import { useReactiveVar } from "@apollo/client";
import { favoriteSongsVar } from "../../apollo/cache";
import { SongData } from "../../utils/types/SongTypes";
import { useMutation } from "@apollo/client";
import { ADD_FAVORITE_SONG, REMOVE_FAVORITE_SONG } from "../../utils/Queries";

type FavoriteProps = {
	song: SongData;
};

const FavoriteButton = ({ song }: FavoriteProps) => {
	const favoriteSongs = useReactiveVar(favoriteSongsVar);
	const isFavorite = favoriteSongs.some((favSong) => favSong.id === song.id);
	const [addFavorite] = useMutation(ADD_FAVORITE_SONG);
	const [removeFavorite] = useMutation(REMOVE_FAVORITE_SONG);

	const handleFavorite = async (e: React.MouseEvent<HTMLButtonElement>) => {
		e.stopPropagation();
		const storedUsername = localStorage.getItem("profileName");
		if (!storedUsername) {
			alert("You need to log in to favorite songs");
			return;
		}
		if (isFavorite) {
			// Remove song from favorites
			const updatedFavorites = favoriteSongs.filter((favoriteSong) => favoriteSong.id !== song.id);
			favoriteSongsVar([...updatedFavorites]);
			await removeFavorite({
				variables: { username: storedUsername, songId: song.id },
			});
		} else {
			// Add song to favorites
			const updatedFavorites = [...favoriteSongs, song];
			favoriteSongsVar([...updatedFavorites]);
			await addFavorite({
				variables: { username: storedUsername, songId: song.id },
			});
		}
	};

	return (
		<button className="favoriteButton" onClick={handleFavorite} type="button">
			{isFavorite ? (
				<FontAwesomeIcon icon={heartSolid} style={{ color: "var(--theme-favorite-pink)" }} />
			) : (
				<FontAwesomeIcon icon={heartRegular} style={{ color: "var(--theme-favorite-pink)" }} />
			)}
		</button>
	);
};

export default FavoriteButton;
