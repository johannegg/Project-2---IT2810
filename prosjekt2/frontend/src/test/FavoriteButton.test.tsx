import React from "react";
import { describe, test, vi, beforeEach, afterEach, expect } from "vitest";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import FavoriteButton from "../components/FavoriteButton/FavoriteButton";
import { favoriteSongsVar } from "../apollo/cache";
import { useReactiveVar, useMutation, ApolloClient, InMemoryCache } from "@apollo/client";
import { ADD_FAVORITE_SONG, REMOVE_FAVORITE_SONG } from "../utils/Queries";
import { faHeart as heartRegular } from "@fortawesome/free-regular-svg-icons";
import { faHeart as heartSolid } from "@fortawesome/free-solid-svg-icons";

// Mock Apollo Client
vi.mock("@apollo/client", async () => {
	const actual = await vi.importActual("@apollo/client");
	return {
		...actual,
		useReactiveVar: vi.fn(),
		useMutation: vi.fn(),
	};
});

vi.mock("@fortawesome/react-fontawesome", () => ({
	FontAwesomeIcon: ({ icon, style }: { icon: any; style: React.CSSProperties }) => {
		// Sjekk `icon` direkte
		const isSolidHeart = icon === heartSolid;
		const isRegularHeart = icon === heartRegular;

		// Sett testId basert p√• ikontypen
		const testId = isSolidHeart ? "icon-fas" : isRegularHeart ? "icon-far" : "unknown";

		return (
			<span data-testid={testId} style={style}>
				Icon
			</span>
		);
	},
}));

// Mock data
const mockSong = {
	id: "1",
	title: "Song One",
	artist: { id: "a1", name: "Artist One" },
	year: 2021,
	genre: { name: "Pop" },
	views: 1000,
	lyrics: "These are the lyrics for Song One.",
};

describe("FavoriteButton", () => {
	const addFavoriteMock = vi.fn();
	const removeFavoriteMock = vi.fn();

	const mockApolloClient = new ApolloClient({
		uri: "http://localhost:4000", // Mock URI
		cache: new InMemoryCache(), // Mock cache
	});

	beforeEach(() => {
		// Set a valid profile name in localStorage
		localStorage.setItem("profileName", "testUser");

		// Mock `useReactiveVar`
		vi.mocked(useReactiveVar).mockReturnValue([]);

		// Mock `useMutation`
		vi.mocked(useMutation).mockImplementation((mutation) => {
			if (mutation === ADD_FAVORITE_SONG) {
				return [
					addFavoriteMock,
					{ loading: false, called: false, client: mockApolloClient, reset: vi.fn() },
				];
			}
			if (mutation === REMOVE_FAVORITE_SONG) {
				return [
					removeFavoriteMock,
					{ loading: false, called: false, client: mockApolloClient, reset: vi.fn() },
				];
			}
			throw new Error("Unknown mutation");
		});
	});

	afterEach(() => {
		cleanup();
		vi.resetAllMocks();
		localStorage.clear();
	});

	// Snapshot Tests
	test("matches snapshot when song is not favorited", () => {
		const { container } = render(<FavoriteButton song={mockSong} />);
		expect(container).toMatchSnapshot();
	});

	test("matches snapshot when song is favorited", () => {
		vi.mocked(useReactiveVar).mockReturnValue([mockSong]);
		const { container } = render(<FavoriteButton song={mockSong} />);
		expect(container).toMatchSnapshot();
	});

	// Functional Tests
	test("renders regular heart icon when song is not favorited", () => {
		render(<FavoriteButton song={mockSong} />);
		expect(screen.getByTestId("icon-far")).toBeInTheDocument();
	});

	test("renders solid heart icon when song is favorited", () => {
		vi.mocked(useReactiveVar).mockReturnValue([mockSong]);
		render(<FavoriteButton song={mockSong} />);
		expect(screen.getByTestId("icon-fas")).toBeInTheDocument();
	});

	test("calls addFavorite mutation when song is not favorited", async () => {
		render(<FavoriteButton song={mockSong} />);
		const button = screen.getByRole("button");
		fireEvent.click(button);

		expect(addFavoriteMock).toHaveBeenCalledWith({
			variables: { username: localStorage.getItem("profileName"), songId: mockSong.id },
		});
	});

	test("calls removeFavorite mutation when song is favorited", async () => {
		vi.mocked(useReactiveVar).mockReturnValue([mockSong]);
		render(<FavoriteButton song={mockSong} />);
		const button = screen.getByRole("button");
		fireEvent.click(button);

		expect(removeFavoriteMock).toHaveBeenCalledWith({
			variables: { username: localStorage.getItem("profileName"), songId: mockSong.id },
		});
	});

	test("shows an alert when user is not logged in and tries to favorite a song", () => {
		const alertMock = vi.spyOn(window, "alert").mockImplementation(() => {});
		localStorage.setItem("profileName", ""); // Simulate not logged in

		render(<FavoriteButton song={mockSong} />);
		const button = screen.getByRole("button");
		fireEvent.click(button);

		expect(alertMock).toHaveBeenCalledWith("You need to log in to favorite songs");
		alertMock.mockRestore();
	});
});
