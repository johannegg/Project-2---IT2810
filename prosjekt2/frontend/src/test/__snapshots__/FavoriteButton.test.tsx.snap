// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoriteButton > matches snapshot when song is favorited 1`] = `
<div>
  <button
    aria-label="Remove Song One from favorites"
    class="favoriteButton"
    type="button"
  >
    <span
      data-testid="icon-fas"
    >
      Icon
    </span>
  </button>
</div>
`;

exports[`FavoriteButton > matches snapshot when song is not favorited 1`] = `
<div>
  <button
    aria-label="Add Song One to favorites"
    class="favoriteButton"
    type="button"
  >
    <span
      data-testid="icon-far"
    >
      Icon
    </span>
  </button>
</div>
`;
